File: server/tools/academic-api.ts
/**
 * Academic APIs - Semantic Scholar & OpenAlex
 * No scraping required, free APIs with open access PDF links
 */

// ============ Types ============

export interface AcademicPaper {
  title: string;
  year: number;
  abstract: string;
  authors: string[];
  venue: string;
  citationCount: number;
  url: string;
  pdfUrl?: string; // Open access PDF if available
  doi?: string;
}

export interface AuthorProfile {
  name: string;
  authorId: string;
  affiliation?: string;
  paperCount: number;
  citationCount: number;
  hIndex?: number;
  papers: AcademicPaper[];
}

// ============ Semantic Scholar API ============
// Docs: https://api.semanticscholar.org/api-docs/

const SEMANTIC_SCHOLAR_API = 'https://api.semanticscholar.org/graph/v1';

/**
 * Search for an author by name on Semantic Scholar
 */
export async function searchAuthorSemanticScholar(
  authorName: string,
  affiliation?: string
): Promise<AuthorProfile | null> {
  try {
    // Search for author
    const searchQuery = affiliation ? `${authorName} ${affiliation}` : authorName;
    const searchUrl = `${SEMANTIC_SCHOLAR_API}/author/search?query=${encodeURIComponent(searchQuery)}&limit=5`;
    
    console.log(`🔍 Semantic Scholar: Searching for "${searchQuery}"...`);
    
    const searchRes = await fetch(searchUrl, {
      headers: { 'Accept': 'application/json' }
    });
    
    if (!searchRes.ok) {
      console.log(`⚠️ Semantic Scholar search failed: ${searchRes.status}`);
      return null;
    }
    
    const searchData = await searchRes.json() as { data: { authorId: string; name: string }[] };
    
    if (!searchData.data || searchData.data.length === 0) {
      console.log(`⚠️ No authors found for "${authorName}"`);
      return null;
    }
    
    // Get first matching author (could improve with affiliation matching)
    const authorId = searchData.data[0].authorId;
    console.log(`✅ Found author ID: ${authorId}`);
    
    // Get author details with papers
    const authorUrl = `${SEMANTIC_SCHOLAR_API}/author/${authorId}?fields=name,affiliations,paperCount,citationCount,hIndex,papers.title,papers.year,papers.abstract,papers.venue,papers.citationCount,papers.url,papers.openAccessPdf,papers.externalIds`;
    
    const authorRes = await fetch(authorUrl, {
      headers: { 'Accept': 'application/json' }
    });
    
    if (!authorRes.ok) {
      console.log(`⚠️ Failed to get author details: ${authorRes.status}`);
      return null;
    }
    
    const authorData = await authorRes.json() as {
      name: string;
      authorId: string;
      affiliations?: string[];
      paperCount: number;
      citationCount: number;
      hIndex?: number;
      papers?: {
        title: string;
        year: number;
        abstract: string;
        venue: string;
        citationCount: number;
        url: string;
        openAccessPdf?: { url: string };
        externalIds?: { DOI?: string };
      }[];
    };
    
    // Transform papers
    const papers: AcademicPaper[] = (authorData.papers || [])
      .filter(p => p.year) // Filter out papers without year
      .sort((a, b) => (b.year || 0) - (a.year || 0)) // Sort by year descending
      .slice(0, 10) // Top 10 recent papers
      .map(p => ({
        title: p.title,
        year: p.year,
        abstract: p.abstract || '',
        authors: [], // Would need separate call to get full author list
        venue: p.venue || '',
        citationCount: p.citationCount || 0,
        url: p.url,
        pdfUrl: p.openAccessPdf?.url,
        doi: p.externalIds?.DOI,
      }));
    
    console.log(`✅ Found ${papers.length} papers, ${papers.filter(p => p.pdfUrl).length} with open access PDFs`);
    
    return {
      name: authorData.name,
      authorId: authorData.authorId,
      affiliation: authorData.affiliations?.[0],
      paperCount: authorData.paperCount,
      citationCount: authorData.citationCount,
      hIndex: authorData.hIndex,
      papers,
    };
  } catch (error) {
    console.error('Semantic Scholar API error:', error);
    return null;
  }
}

// ============ OpenAlex API ============
// Docs: https://docs.openalex.org/

const OPENALEX_API = 'https://api.openalex.org';

/**
 * Search for an author by name on OpenAlex (fallback)
 */
export async function searchAuthorOpenAlex(
  authorName: string,
  affiliation?: string
): Promise<AuthorProfile | null> {
  try {
    // Search for author
    const searchUrl = `${OPENALEX_API}/authors?search=${encodeURIComponent(authorName)}&per_page=5`;
    
    console.log(`🔍 OpenAlex: Searching for "${authorName}"...`);
    
    const searchRes = await fetch(searchUrl, {
      headers: { 
        'Accept': 'application/json',
        'User-Agent': 'PhDApply/1.0 (mailto:contact@example.com)' // OpenAlex asks for this
      }
    });
    
    if (!searchRes.ok) {
      console.log(`⚠️ OpenAlex search failed: ${searchRes.status}`);
      return null;
    }
    
    const searchData = await searchRes.json() as { 
      results: { 
        id: string; 
        display_name: string;
        last_known_institution?: { display_name: string };
        works_count: number;
        cited_by_count: number;
        summary_stats?: { h_index: number };
      }[] 
    };
    
    if (!searchData.results || searchData.results.length === 0) {
      console.log(`⚠️ No authors found on OpenAlex`);
      return null;
    }
    
    // Find best match (optionally by affiliation)
    let author = searchData.results[0];
    if (affiliation) {
      const affiliationLower = affiliation.toLowerCase();
      const matchingAuthor = searchData.results.find(a => 
        a.last_known_institution?.display_name?.toLowerCase().includes(affiliationLower)
      );
      if (matchingAuthor) author = matchingAuthor;
    }
    
    console.log(`✅ Found author: ${author.display_name} (${author.last_known_institution?.display_name || 'Unknown affiliation'})`);
    
    // Get author's works
    const worksUrl = `${OPENALEX_API}/works?filter=author.id:${author.id}&sort=publication_date:desc&per_page=10`;
    
    const worksRes = await fetch(worksUrl, {
      headers: { 
        'Accept': 'application/json',
        'User-Agent': 'PhDApply/1.0'
      }
    });
    
    if (!worksRes.ok) {
      console.log(`⚠️ Failed to get works: ${worksRes.status}`);
      return {
        name: author.display_name,
        authorId: author.id,
        affiliation: author.last_known_institution?.display_name,
        paperCount: author.works_count,
        citationCount: author.cited_by_count,
        hIndex: author.summary_stats?.h_index,
        papers: [],
      };
    }
    
    const worksData = await worksRes.json() as {
      results: {
        title: string;
        publication_year: number;
        abstract_inverted_index?: Record<string, number[]>;
        primary_location?: { source?: { display_name: string } };
        cited_by_count: number;
        doi?: string;
        open_access?: { oa_url?: string };
        authorships?: { author: { display_name: string } }[];
      }[];
    };
    
    // Transform papers
    const papers: AcademicPaper[] = worksData.results.map(w => ({
      title: w.title,
      year: w.publication_year,
      abstract: invertedIndexToText(w.abstract_inverted_index),
      authors: w.authorships?.map(a => a.author.display_name) || [],
      venue: w.primary_location?.source?.display_name || '',
      citationCount: w.cited_by_count,
      url: w.doi ? `https://doi.org/${w.doi}` : '',
      pdfUrl: w.open_access?.oa_url,
      doi: w.doi,
    }));
    
    console.log(`✅ Found ${papers.length} papers, ${papers.filter(p => p.pdfUrl).length} with open access`);
    
    return {
      name: author.display_name,
      authorId: author.id,
      affiliation: author.last_known_institution?.display_name,
      paperCount: author.works_count,
      citationCount: author.cited_by_count,
      hIndex: author.summary_stats?.h_index,
      papers,
    };
  } catch (error) {
    console.error('OpenAlex API error:', error);
    return null;
  }
}

/**
 * Convert OpenAlex inverted index abstract to text
 */
function invertedIndexToText(invertedIndex?: Record<string, number[]>): string {
  if (!invertedIndex) return '';
  
  const words: [string, number][] = [];
  for (const [word, positions] of Object.entries(invertedIndex)) {
    for (const pos of positions) {
      words.push([word, pos]);
    }
  }
  
  words.sort((a, b) => a[1] - b[1]);
  return words.map(w => w[0]).join(' ');
}

// ============ Combined Search ============

/**
 * Search for author using Semantic Scholar first, then OpenAlex as fallback
 */
export async function searchAuthor(
  authorName: string, 
  affiliation?: string
): Promise<AuthorProfile | null> {
  // Try Semantic Scholar first (better paper data)
  let result = await searchAuthorSemanticScholar(authorName, affiliation);
  
  // Fallback to OpenAlex if no results
  if (!result || result.papers.length === 0) {
    console.log('📚 Trying OpenAlex as fallback...');
    result = await searchAuthorOpenAlex(authorName, affiliation);
  }
  
  return result;
}

/**
 * Download PDF from open access URL
 */
export async function downloadOpenAccessPdf(url: string): Promise<Uint8Array | null> {
  try {
    console.log(`📥 Downloading open access PDF: ${url}`);
    
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
      },
    });
    
    if (!response.ok) {
      console.log(`⚠️ PDF download failed: ${response.status}`);
      return null;
    }
    
    const contentType = response.headers.get('content-type') || '';
    if (!contentType.includes('application/pdf')) {
      console.log(`⚠️ Not a PDF: ${contentType}`);
      return null;
    }
    
    const buffer = await response.arrayBuffer();
    if (buffer.byteLength < 10000) {
      console.log(`⚠️ PDF too small: ${buffer.byteLength} bytes`);
      return null;
    }
    
    console.log(`✅ Downloaded PDF: ${buffer.byteLength} bytes`);
    return new Uint8Array(buffer);
  } catch (error) {
    console.log(`⚠️ PDF download error: ${error}`);
    return null;
  }
}

File: autogen_service/app/tools/academic_api.py
from __future__ import annotations

from typing import Any
from urllib.parse import quote

import httpx

OPENALEX_API = "https://api.openalex.org"
DEFAULT_USER_AGENT = "PhDApply-AutoGen/1.0 (mailto:contact@example.com)"


def _inverted_index_to_text(inverted_index: dict[str, list[int]] | None) -> str:
    if not inverted_index:
        return ""

    words: list[tuple[str, int]] = []
    for word, positions in inverted_index.items():
        for pos in positions:
            words.append((word, pos))
    words.sort(key=lambda item: item[1])
    return " ".join(word for word, _ in words)


def _is_valid_pdf_url(url: str | None) -> bool:
    if not url:
        return False
    if not url.startswith("http"):
        return False
    lower = url.lower()
    if lower.endswith(".pdf"):
        return True
    if "arxiv.org/pdf/" in lower:
        return True
    if "/pdf/" in lower or "/pdfs/" in lower:
        return True
    return False


def _normalize_work(work: dict[str, Any]) -> dict[str, Any]:
    pdf_url: str | None = None
    best_oa_location = work.get("best_oa_location") or {}
    primary_location = work.get("primary_location") or {}
    locations = work.get("locations") or []

    if _is_valid_pdf_url(best_oa_location.get("pdf_url")):
        pdf_url = best_oa_location.get("pdf_url")
    elif _is_valid_pdf_url(primary_location.get("pdf_url")):
        pdf_url = primary_location.get("pdf_url")
    else:
        for location in locations:
            candidate = location.get("pdf_url")
            if _is_valid_pdf_url(candidate):
                pdf_url = candidate
                break

    doi = work.get("doi")
    return {
        "title": work.get("title") or "",
        "year": work.get("publication_year") or 0,
        "abstract": _inverted_index_to_text(work.get("abstract_inverted_index")),
        "authors": [
            a.get("author", {}).get("display_name", "")
            for a in (work.get("authorships") or [])
            if a.get("author", {}).get("display_name")
        ],
        "venue": (primary_location.get("source") or {}).get("display_name", ""),
        "citationCount": work.get("cited_by_count") or 0,
        "url": f"https://doi.org/{doi}" if doi else "",
        "pdfUrl": pdf_url,
        "doi": doi,
    }


async def search_author_openalex(
    author_name: str, affiliation: str | None = None
) -> dict[str, Any] | None:
    params = {"search": author_name, "per_page": 10}
    headers = {"Accept": "application/json", "User-Agent": DEFAULT_USER_AGENT}

    async with httpx.AsyncClient(timeout=25.0) as client:
        search_resp = await client.get(f"{OPENALEX_API}/authors", params=params, headers=headers)
        if search_resp.status_code != 200:
            return None

        search_data = search_resp.json()
        candidates = search_data.get("results") or []
        if not candidates:
            return None

        selected = candidates[0]
        if affiliation:
            affiliation_lower = affiliation.lower()
            for candidate in candidates:
                institutions = candidate.get("last_known_institutions") or []
                if any(
                    (inst.get("display_name") or "").lower().find(affiliation_lower) >= 0
                    for inst in institutions
                ):
                    selected = candidate
                    break
                last_known = candidate.get("last_known_institution") or {}
                if affiliation_lower in (last_known.get("display_name") or "").lower():
                    selected = candidate
                    break

        author_id = selected.get("id")
        if not author_id:
            return None

        recent_url = f"{OPENALEX_API}/works"
        recent_params = {
            "filter": f"author.id:{author_id}",
            "sort": "publication_date:desc",
            "per_page": 12,
        }
        cited_params = {
            "filter": f"author.id:{author_id}",
            "sort": "cited_by_count:desc",
            "per_page": 12,
        }

        recent_resp, cited_resp = await client.get(
            recent_url, params=recent_params, headers=headers
        ), await client.get(recent_url, params=cited_params, headers=headers)

        recent_results = (
            (recent_resp.json().get("results") or []) if recent_resp.status_code == 200 else []
        )
        cited_results = (
            (cited_resp.json().get("results") or []) if cited_resp.status_code == 200 else []
        )

    merged: list[dict[str, Any]] = []
    seen_titles: set[str] = set()
    for work in [*recent_results, *cited_results]:
        normalized = _normalize_work(work)
        title_key = normalized["title"].strip().lower()
        if not title_key or title_key in seen_titles:
            continue
        seen_titles.add(title_key)
        merged.append(normalized)

    merged = merged[:12]

    last_known_inst = selected.get("last_known_institution") or {}
    institutions = selected.get("last_known_institutions") or []
    affiliation_name = (
        last_known_inst.get("display_name")
        or (institutions[0].get("display_name") if institutions else None)
        or affiliation
        or "Unknown"
    )

    return {
        "name": selected.get("display_name") or author_name,
        "authorId": author_id,
        "affiliation": affiliation_name,
        "paperCount": selected.get("works_count") or len(merged),
        "citationCount": selected.get("cited_by_count") or 0,
        "hIndex": (selected.get("summary_stats") or {}).get("h_index"),
        "papers": merged,
        "source": f"{OPENALEX_API}/authors/{quote(author_id, safe=':/')}",
    }


async def search_paper_by_title(
    title: str, keywords: list[str] | None = None
) -> dict[str, Any] | None:
    query = title
    if keywords:
        query = f"{title} {' '.join(keywords)}"

    params = {"search": query, "per_page": 8}
    headers = {"Accept": "application/json", "User-Agent": DEFAULT_USER_AGENT}

    async with httpx.AsyncClient(timeout=20.0) as client:
        response = await client.get(f"{OPENALEX_API}/works", params=params, headers=headers)
        if response.status_code != 200:
            return None
        works = response.json().get("results") or []

    title_words = [w for w in title.lower().split() if len(w) > 3]
    if not title_words:
        title_words = title.lower().split()

    best_candidate: dict[str, Any] | None = None
    best_score = 0.0
    for work in works:
        normalized = _normalize_work(work)
        if not normalized.get("pdfUrl"):
            continue

        candidate_title = (normalized.get("title") or "").lower()
        if not candidate_title:
            continue
        overlap = sum(1 for w in title_words if w in candidate_title)
        score = overlap / max(len(title_words), 1)
        if score > best_score:
            best_score = score
            best_candidate = normalized

    if best_candidate and best_score >= 0.45:
        return best_candidate
    return None
